// @sudoku/stores/gameDomain.js
import { writable, derived } from 'svelte/store';
import { createSudoku, createGame, createSudokuFromJSON, createGameFromJSON } from '../../../domain/index.js';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { decodeSencode } from '@sudoku/sencode';

function createGameStore() {
  const state = writable({
    grid: null,
    givens: null,
    invalidCells: [],
    canUndo: false,
    canRedo: false,
    isExploring: false,
    hasConflicts: false,
    isComplete: false,
    game: null
  });

  let gameInstance = null;

  function updateState() {
    if (gameInstance) {
      state.set({
        grid: gameInstance.getGrid(),
        givens: gameInstance.getGivens(),
        invalidCells: gameInstance.getInvalidCells(),
        canUndo: gameInstance.canUndo(),
        canRedo: gameInstance.canRedo(),
        isExploring: gameInstance.isExploring(),
        hasConflicts: gameInstance.hasConflicts(),
        isComplete: gameInstance.isComplete(),
        game: gameInstance
      });
    }
  }

  const store = {
    subscribe: state.subscribe,

    startNew(difficulty) {
      const gridData = generateSudoku(difficulty);
      const sudoku = createSudoku(gridData);
      gameInstance = createGame({ sudoku });
      updateState();
    },

    startCustom(sencode) {
      const gridData = decodeSencode(sencode);
      const sudoku = createSudoku(gridData);
      gameInstance = createGame({ sudoku });
      updateState();
    },

    guess(pos, value) {
      if (!gameInstance) return false;
      const { x, y } = pos;
      const success = gameInstance.guess({ row: y, col: x, value });
      updateState();
      return success;
    },

    clear(pos) {
      if (!gameInstance) return false;
      const { x, y } = pos;
      if (gameInstance.isGiven(y, x)) return false;
      gameInstance.guess({ row: y, col: x, value: 0 });
      updateState();
      return true;
    },

    undo() {
      if (!gameInstance) return;
      gameInstance.undo();
      updateState();
    },

    redo() {
      if (!gameInstance) return;
      gameInstance.redo();
      updateState();
    },

    getHint() {
      if (!gameInstance) return null;
      return gameInstance.getHint();
    },

    getCandidates(pos) {
      if (!gameInstance) return [];
      const { x, y } = pos;
      return gameInstance.getCandidates(y, x);
    },

    isGiven(pos) {
      if (!gameInstance) return false;
      const { x, y } = pos;
      return gameInstance.isGiven(y, x);
    },

    isExploring() {
      if (!gameInstance) return false;
      return gameInstance.isExploring();
    },

    enterExplore() {
      if (!gameInstance) return false;
      const result = gameInstance.enterExplore();
      updateState();
      return result;
    },

    commitExploration() {
      if (!gameInstance) return false;
      const result = gameInstance.commitExploration();
      updateState();
      return result;
    },

    abandonExploration() {
      if (!gameInstance) return false;
      const result = gameInstance.abandonExploration();
      updateState();
      return result;
    },

    isComplete() {
      if (!gameInstance) return false;
      return gameInstance.isComplete();
    },

    applyHint(pos) {
      if (!gameInstance) return false;
      const { x, y } = pos;
      const currentGrid = gameInstance.getGrid();
      const solvedGrid = solveSudoku(currentGrid);
      const value = solvedGrid[y][x];
      if (value === 0) return false;
      const success = gameInstance.guess({ row: y, col: x, value });
      updateState();
      return success;
    },

    toJSON() {
      if (!gameInstance) return null;
      return gameInstance.toJSON();
    },

    loadFromJSON(json) {
      gameInstance = createGameFromJSON(json);
      updateState();
    }
  };

  return store;
}

export const gameDomain = createGameStore();
